### PR TITLE
[Az.DataProtection] Added soft delete and MUA feature for Backup vaults

### PR DESCRIPTION
**Pull request by hiaga**
## Description

 Fixed an issue in New-AzDataProtectionBackupConfigurationClientObject cmdlet
 Added AKS Hooks
 Added soft delete and MUA features
 Added test case

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/DataProtection/examples/Get-AzDataProtectionResourceGuardMapping.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
```
Comment (YanaXu) [inline src/DataProtection/examples/Get-AzDataProtectionResourceGuardMapping.md:3]: Add at least one example.

File: src/DataProtection/docs/Get-AzDataProtectionResourceGuardMapping.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+```
+
+## DESCRIPTION
+Returns the ResourceGuardProxy object associated with the vault, and that matches the name in the request
+
+## EXAMPLES
+
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
```
Comment (YanaXu) [inline src/DataProtection/docs/Get-AzDataProtectionResourceGuardMapping.md:40]: add examples and run build-module.ps1 again.

File: src/DataProtection/examples/Get-AzDataProtectionSoftDeletedBackupInstance.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
```
Comment (YanaXu) [inline src/DataProtection/examples/Get-AzDataProtectionSoftDeletedBackupInstance.md:3]: Add examples.

File: src/DataProtection/examples/Get-AzDataProtectionResourceGuardMapping.md (hunk 1 above)
Comment (hiaga) [reply src/DataProtection/examples/Get-AzDataProtectionResourceGuardMapping.md]: done

**Comment by hiaga** [comment]
\azp run

**Comment by YanaXu** [comment]
Hi @hiaga , please check CI failures and fix them.

**Comment by hiaga** [comment]
> Hi @hiaga , please check CI failures and fix them.

Updated

**Comment by BethanyZhou** [comment]
Hi @hiaga, please double confirm issues reported by CI analyzer. If they are false positive, please suppress them by following 
```
"Module","Cmdlet","Example","Line","RuleName","ProblemId","Severity","Description","Extent","Remediation"
  "Az.DataProtection","New-AzDataProtectionBackupConfigurationClientObject","2","1","Invalid_Cmdlet","5000","1","Get-AzStorageAccount is not a valid command name.","Get-AzStorageAccount -ResourceGroupName $resourceGroupName -Name $storageAccountName","Check the spell of Get-AzStorageAccount."
  "Az.DataProtection","New-AzDataProtectionBackupConfigurationClientObject","2","2","Invalid_Cmdlet","5000","1","Get-AzStorageContainer is not a valid command name.","Get-AzStorageContainer -Context $storageAccount.Context","Check the spell of Get-AzStorageContainer."
 ```

**Comment by hiaga** [comment]
> Hi @hiaga, please double confirm issues reported by CI analyzer. If they are false positive, please suppress them by following https://github.com/Azure/azure-powershell/blob/main/documentation/tooling/static-analysis.md#suppressing-exceptions
> 
> ```
> "Module","Cmdlet","Example","Line","RuleName","ProblemId","Severity","Description","Extent","Remediation"
>   "Az.DataProtection","New-AzDataProtectionBackupConfigurationClientObject","2","1","Invalid_Cmdlet","5000","1","Get-AzStorageAccount is not a valid command name.","Get-AzStorageAccount -ResourceGroupName $resourceGroupName -Name $storageAccountName","Check the spell of Get-AzStorageAccount."
>   "Az.DataProtection","New-AzDataProtectionBackupConfigurationClientObject","2","2","Invalid_Cmdlet","5000","1","Get-AzStorageContainer is not a valid command name.","Get-AzStorageContainer -Context $storageAccount.Context","Check the spell of Get-AzStorageContainer."
> ```

@BethanyZhou Confirming that the issues flagged are false positive. Suppressed as suggested.

File: src/DataProtection/custom/Cmdlets/Platform/Restore/New-AzDataProtectionRestoreConfigurationClientObject.ps1 [hunk 4]
```diff
@@ -40,17 +40,21 @@ function New-AzDataProtectionRestoreConfigurationClientObject{
...
         ${NamespaceMapping},
 
         [Parameter(Mandatory=$false, HelpMessage='Restore mode for persistent volumes. Allowed values are RestoreWithVolumeData, RestoreWithoutVolumeData. Default value is RestoreWithVolumeData')]
         [System.String]
         [ValidateSet('RestoreWithVolumeData','RestoreWithoutVolumeData')]
-        ${PersistentVolumeRestoreMode}
+        ${PersistentVolumeRestoreMode},
+        
+        [Parameter(Mandatory=$false, HelpMessage='Hook reference to be executed during restore.')]
+        [Microsoft.Azure.PowerShell.Cmdlets.DataProtection.Models.Api20230501.NamespacedNameResource[]]
```
Comment (BethanyZhou) [inline src/DataProtection/custom/Cmdlets/Platform/Restore/New-AzDataProtectionRestoreConfigurationClientObject.ps1:52]: Please confirm whether a helper comdlet is needed for RestoreHookReference NamespacedNameResource[]

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/DataProtection/custom/Cmdlets/Platform/Restore/New-AzDataProtectionRestoreConfigurationClientObject.ps1 (hunk 4 above)
Comment (hiaga) [reply src/DataProtection/custom/Cmdlets/Platform/Restore/New-AzDataProtectionRestoreConfigurationClientObject.ps1:52]: Hi @BethanyZhou, it's a list of an object with two fields. we wouldn't need a separate command for now. However, an example would help, let me add an example.

**Review by BethanyZhou — APPROVED** [review]
Looks good to me

**Pull request merged**